src/NuGet.Core/NuGet.Common/ConcurrencyUtilities.cs

```diff
@@ -2,6 +2,7 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
+using System.Collections.Generic;
```

Comment:
Look into the sync locking. 

Also consider a class for the semaphores themselves.